src/org/skypro/skyshop/product/ProductBasket.java

```diff
@@ -8,99 +8,108 @@
 import javax.imageio.metadata.IIOInvalidTreeException;
 
 public class ProductBasket {
-    private LinkedList<Product> array;
+    private Map<String, List<Product>> array;
     private int productsCount = 0;
 
     public ProductBasket() {
-        this.array = new LinkedList<>();
+        this.array = new HashMap<String, List<Product>>();
     }
 
     public void addProductToBasket(Product product) {
-        this.array.add(product);
-        this.productsCount += 1;
+        List<Product> products = array.get(product.getNameOfProduct());
+
+        if (products == null) {
+            products = new ArrayList<>();
+        }
+
+        products.add(product);
+
+        array.put(product.nameOfProduct, products);
 
+        this.productsCount += 1;
     }
 
     public List<Product> removeProductFromBasket(String name) {
-        List<Product> removedProducts = new ArrayList<>();
-        Iterator<Product> iterator = array.iterator();
-
-        while (iterator.hasNext()) {
-            Product product = iterator.next();
-            if (product != null && product.getNameOfProduct().equals(name)) {
-                removedProducts.add(product);
-                iterator.remove();
-                productsCount -= 1;
-            }
-            if (product.equals(null)) {
-                System.out.println("Список пуст");
-            }
+        List<Product> removedProducts = array.get(name);
+
+        if (removedProducts != null) {
+            productsCount = removedProducts.size();
+            array.remove(name);
+            return removedProducts;
         }
-        return removedProducts;
+
+        System.out.println("Товар с именем " + name + " не найден");
+        return new ArrayList<>();
```

Comment:
лучше вернуть пустой список, чем список из 10 null элементов

```suggestion
        return Collections.emptyList();
```